File: routes/movimientos.py
from flask import Blueprint, request, jsonify
from flask_mysqldb import MySQL

movimientos_bp = Blueprint(
    'movimientos',
    __name__
)


@movimientos_bp.route('/', methods=['GET'])
def obtener_movimientos():

    cursor = mysql.connection.cursor()

    cursor.execute("""
        SELECT
            m.id_movimiento,
            m.fecha,
            p.codigo,
            p.nombre,
            c.codigo_qr,
            m.tipo,
            m.cantidad,
            m.stock_anterior,
            m.stock_nuevo,
            m.motivo
        FROM movimientos m
        INNER JOIN productos p
            ON m.id_producto = p.id_producto
        LEFT JOIN cajas c
            ON m.id_caja = c.id_caja
        ORDER BY m.fecha DESC
    """)

    movimientos = cursor.fetchall()

    cursor.close()

    resultado = []

    for movimiento in movimientos:

        resultado.append({
            'id_movimiento': movimiento[0],
            'fecha': movimiento[1].isoformat()
                if movimiento[1] else None,
            'codigo_producto': movimiento[2],
            'producto': movimiento[3],
            'codigo_caja': movimiento[4],
            'tipo': movimiento[5],
            'cantidad': movimiento[6],
            'stock_anterior': movimiento[7],
            'stock_nuevo': movimiento[8],
            'motivo': movimiento[9]
        })

    return jsonify(resultado)


@movimientos_bp.route('/', methods=['POST'])
def crear_movimiento():

    datos = request.get_json()

    id_producto = datos.get('id_producto')
    id_caja = datos.get('id_caja')
    tipo = datos.get('tipo')
    cantidad = datos.get('cantidad')
    motivo = datos.get('motivo')

    if not id_producto or not tipo or not cantidad:
        return jsonify({
            'error': 'Producto, tipo y cantidad son obligatorios'
        }), 400

    if tipo not in ['ENTRADA', 'SALIDA', 'AJUSTE']:
        return jsonify({
            'error': 'Tipo de movimiento inválido'
        }), 400

    cursor = mysql.connection.cursor()

    try:

        cursor.execute("""
            SELECT stock
            FROM productos
            WHERE id_producto = %s
            FOR UPDATE
        """, (id_producto,))

        producto = cursor.fetchone()

        if not producto:
            mysql.connection.rollback()

            return jsonify({
                'error': 'Producto no encontrado'
            }), 404

        stock_anterior = producto[0]

        if tipo == 'ENTRADA':
            stock_nuevo = stock_anterior + cantidad

        elif tipo == 'SALIDA':
            stock_nuevo = stock_anterior - cantidad

            if stock_nuevo < 0:
                mysql.connection.rollback()

                return jsonify({
                    'error': 'No hay suficiente stock'
                }), 400

        else:
            stock_nuevo = cantidad

        cursor.execute("""
            UPDATE productos
            SET stock = %s
            WHERE id_producto = %s
        """, (
            stock_nuevo,
            id_producto
        ))

        cursor.execute("""
            INSERT INTO movimientos
            (
                id_producto,
                id_caja,
                tipo,
                cantidad,
                stock_anterior,
                stock_nuevo,
                motivo
            )
            VALUES (%s, %s, %s, %s, %s, %s, %s)
        """, (
            id_producto,
            id_caja,
            tipo,
            cantidad,
            stock_anterior,
            stock_nuevo,
            motivo
        ))

        mysql.connection.commit()

        return jsonify({
            'mensaje': 'Movimiento registrado correctamente',
            'stock_anterior': stock_anterior,
            'stock_nuevo': stock_nuevo
        }), 201

    except Exception as e:

        mysql.connection.rollback()

        return jsonify({
            'error': str(e)
        }), 500

    finally:
        cursor.close()


@movimientos_bp.route('/recientes', methods=['GET'])
def movimientos_recientes():

    cursor = mysql.connection.cursor()

    cursor.execute("""
        SELECT
            m.id_movimiento,
            m.fecha,
            p.nombre,
            c.codigo_qr,
            m.tipo,
            m.cantidad,
            m.stock_nuevo
        FROM movimientos m
        INNER JOIN productos p
            ON m.id_producto = p.id_producto
        LEFT JOIN cajas c
            ON m.id_caja = c.id_caja
        ORDER BY m.fecha DESC
        LIMIT 10
    """)

    movimientos = cursor.fetchall()

    cursor.close()

    resultado = []

    for movimiento in movimientos:

        resultado.append({
            'id_movimiento': movimiento[0],
            'fecha': movimiento[1].isoformat()
                if movimiento[1] else None,
            'producto': movimiento[2],
            'caja': movimiento[3],
            'tipo': movimiento[4],
            'cantidad': movimiento[5],
            'stock_nuevo': movimiento[6]
        })

    return jsonify(resultado)

File: static/js/movimientos.js
document.addEventListener("DOMContentLoaded", () => {
    cargarMovimientos();
    cargarProductosSelect();
    cargarCajasSelect();

    const form = document.getElementById("formMovimiento");
    if (form) {
        form.addEventListener("submit", guardarMovimiento);
    }
});

// 1. Cargar la tabla de movimientos desde la API
async function cargarMovimientos() {
    const tbody = document.getElementById("tablaMovimientos");
    if (!tbody) return;

    try {
        const respuesta = await fetch("/api/movimientos/");
        if (!respuesta.ok) throw new Error("Error al obtener los movimientos");

        const movimientos = await respuesta.json();

        if (movimientos.length === 0) {
            tbody.innerHTML = `<tr><td colspan="7" class="empty">No hay movimientos registrados.</td></tr>`;
            return;
        }

        tbody.innerHTML = movimientos.map(m => {
            const fechaFormateada = m.fecha ? new Date(m.fecha).toLocaleString() : "-";
            const cajaNombre = m.codigo_caja ? `Caja: ${m.codigo_caja}` : "-";
            
            let badgeColor = "#17a2b8";
            if (m.tipo === "ENTRADA") badgeColor = "#28a745";
            if (m.tipo === "SALIDA") badgeColor = "#dc3545";
            if (m.tipo === "AJUSTE") badgeColor = "#ffc107";

            return `
                <tr>
                    <td>${fechaFormateada}</td>
                    <td><strong>${m.producto || "-"}</strong> <br><small style="color:#666">${m.codigo_producto || ""}</small></td>
                    <td>${cajaNombre}</td>
                    <td><span style="background:${badgeColor}; color:white; padding: 2px 8px; border-radius: 4px; font-size:0.85em; font-weight:bold;">${m.tipo}</span></td>
                    <td><strong>${m.cantidad}</strong></td>
                    <td>${m.stock_anterior} ➔ <strong>${m.stock_nuevo}</strong></td>
                    <td>${m.motivo || "-"}</td>
                </tr>
            `;
        }).join("");

    } catch (error) {
        console.error("Error al cargar movimientos:", error);
        tbody.innerHTML = `<tr><td colspan="7" class="empty" style="color:red;">Error al cargar datos.</td></tr>`;
    }
}

// 2. Cargar productos en el select del formulario
async function cargarProductosSelect() {
    const select = document.getElementById("id_producto");
    if (!select) return;

    try {
        const respuesta = await fetch("/api/productos/");
        if (!respuesta.ok) return;

        const productos = await respuesta.json();
        select.innerHTML = `<option value="">Selecciona un producto</option>` +
            productos.map(p => `<option value="${p.id_producto}">${p.nombre} (${p.codigo || 'Sin código'}) - Stock: ${p.stock}</option>`).join("");
    } catch (error) {
        console.error("Error al cargar productos:", error);
    }
}

// 3. Cargar cajas en el select del formulario
async function cargarCajasSelect() {
    const select = document.getElementById("id_caja");
    if (!select) return;

    try {
        const respuesta = await fetch("/api/cajas");
        if (!respuesta.ok) return;

        const cajas = await respuesta.json();
        select.innerHTML = `<option value="">Sin caja asignada</option>` +
            cajas.map(c => `<option value="${c.id_caja}">${c.nombre} (${c.codigo_qr || 'Sin QR'})</option>`).join("");
    } catch (error) {
        console.error("Error al cargar cajas:", error);
    }
}

// 4. Guardar un nuevo movimiento (POST)
async function guardarMovimiento(e) {
    e.preventDefault();

    const id_producto = document.getElementById("id_producto").value;
    const tipo = document.getElementById("tipo").value;
    const cantidad = parseInt(document.getElementById("cantidad").value);
    const id_caja_val = document.getElementById("id_caja").value;
    const id_caja = id_caja_val ? parseInt(id_caja_val) : null;
    const motivo = document.getElementById("motivo").value.trim();

    if (!id_producto || !tipo || !cantidad) {
        Swal.fire("Atención", "Producto, tipo y cantidad son obligatorios", "warning");
        return;
    }

    try {
        const respuesta = await fetch("/api/movimientos/", {
            method: "POST",
            headers: { "Content-Type": "application/json" },
            body: JSON.stringify({
                id_producto: parseInt(id_producto),
                id_caja: id_caja,
                tipo: tipo,
                cantidad: cantidad,
                motivo: motivo
            })
        });

        const resultado = await respuesta.json();

        if (respuesta.ok) {
            Swal.fire("¡Éxito!", resultado.mensaje || "Movimiento registrado", "success");
            cerrarModal();
            cargarMovimientos();
            cargarProductosSelect(); // Actualiza stocks en el dropdown
        } else {
            Swal.fire("Error", resultado.error || "No se pudo registrar el movimiento", "error");
        }
    } catch (error) {
        console.error("Error al guardar movimiento:", error);
        Swal.fire("Error", "Ocurrió un fallo en el servidor", "error");
    }
}

// 5. Controles de Modal
function abrirModalNuevo() {
    const modal = document.getElementById("modalMovimiento");
    const form = document.getElementById("formMovimiento");

    if (form) form.reset();
    if (modal) {
        modal.style.display = "flex";
        modal.classList.remove("hidden");
    }
}

function cerrarModal() {
    const modal = document.getElementById("modalMovimiento");
    if (modal) {
        modal.style.display = "none";
        modal.classList.add("hidden");
    }
}